compiler/parser/RdrHsSyn.hs

```diff
@@ -2437,28 +2437,57 @@ mkRdrPat = VarPat noExt . noLoc
 mkVarPat :: OccName -> Pat GhcPs
 mkVarPat = VarPat noExt . noLoc . mkRdrUnqual
 
--- | Utility for constructing patterns of the form 'arg@T'.
-asPatPrefixCon :: String -> Located RdrName -> Pat GhcPs
-asPatPrefixCon varName conName =
-  AsPat noExt
-    (noLoc $ mkRdrUnqual (mkVarOcc varName))
-    (noLoc $ ConPatIn conName $ PrefixCon [])
-
--- | Utility for constructing patterns of the form 'arg@T{..}'.
-asPatRecWild :: String -> Located RdrName -> Pat GhcPs
-asPatRecWild varName conName =
-  AsPat noExt
-    (noLoc $ mkRdrUnqual (mkVarOcc varName))
-    (noLoc $ ConPatIn conName $
-      RecCon $ HsRecFields { rec_flds = [], rec_dotdot = Just 0 })
-
--- | Utility function for constructing patterns of the form 'arg@X'
--- where 'X' is either a record wildcard pattern or a prefix
--- constructor pattern depending on whether a non-empty record field
--- list is provided.
-argPatOfChoice :: Located RdrName -> LHsType GhcPs -> Pat GhcPs
-argPatOfChoice choiceConName (L _ (HsRecTy _ [])) = asPatPrefixCon "arg" choiceConName
-argPatOfChoice choiceConName _ = asPatRecWild "arg" choiceConName
+-- | Utility for constructing patterns of the form 'arg@T{..}',
+-- omitting the '..' for types without labelled fields, and
+-- omitting the 'arg' binding for types with a field of the same name.
+asPatRecWild :: String -> Located RdrName -> LHsType GhcPs -> Pat GhcPs
+asPatRecWild varName conName conType =
+  asPatRecWild' varName conName (recFieldNames conType)
+
+-- | Like 'asPatRecWild', but takes the field names as a list of strings
+-- instead of extracting them from an 'LHsType GhcPs'
+asPatRecWild' :: String -> Located RdrName -> [String] -> Pat GhcPs
+asPatRecWild' varName conName fieldNames = do
+  asPat $ patRecWild' conName fieldNames
+  where
+    asPat =
+      if varName `elem` fieldNames then
+        id
+      else
+        AsPat noExt (noLoc $ mkRdrUnqual $ mkVarOcc varName) . noLoc
+
+-- | Utility for constructing patterns of the form 'T{..}',
+-- omitting the '..' for types without labelled fields.
```

Comment:
this is a bit better than using `PrefixCon` for types without fields because `{}` also works for non-record product types, i.e.

```
data Prod = Prod Int Char

f : Prod -> Bool
f Prod {} = True
```